### PR TITLE
System tests + complete client implementation

### PR DESCRIPTION
**Pull request by brstgt**
Hi there,

I want to use the async client in our project. But unfortunately it is missing background jobs + unique ids.
I took the occasion and completed the implementation and wrote system tests for it, which revealed some edge cases and failures.

To make the tests easier to write and to understand, I used AMPHP. Unfortunately this is only works with PHP 7.
My 2 cents: I'd like to keep it. Code is so much more beautiful and better to understand than with the react promise hell. I'd even go so far to drop support for 5.6 as it isn't even actively supported any more.

WDYT?

To run the system tests on travis, it still needs some tweaks.

### CODE REVIEW

**Comment by bzikarsky** [comment]
First of all: This is awesome! (And a lof of code to review ;-))

I think this is pretty much mergeable (when the conflicts are resolved). The public API is even staying more or less stable. 

Some minor ideas:
1) The testing groups may profit from a little explanation in the README.
2) submit and submitBackground are very similiar. I'm not so sure if we can reduce some of the redundancies which are not related to the task being a background or a normal task.

File: src/Protocol/Connection.php [hunk 1]
```diff
@@ -40,7 +41,7 @@ class Connection extends EventEmitter
     /**
      * @var Stream
      */
-    protected $stream;
+    public $stream;
```
Comment (bzikarsky) [inline src/Protocol/Connection.php]: What is the reason for making `$stream` public? The tests?

Comment (brstgt) [reply src/Protocol/Connection.php]: It is required to pause the connection if there are no pending tasks. This enables the loop to end without explicitly disconnecting the client. Other options: add pause/resume method to connection or expose the stream via a getter

Comment (bzikarsky) [reply src/Protocol/Connection.php]: I'd suggest pause/resume methods. This is still a kind of leaky abstraction but better than full access to the connection object.
